### PR TITLE
Add rate limit to signup

### PR DESCRIPTION
**Pull request by openprojects**
Add rate limit to Sign in to control the number of Verification Emails sent per day.

### CODE REVIEW

File: indico/modules/auth/controllers.py [hunk 1]
```diff
@@ -273,7 +274,15 @@ def _process(self):
     def _process_verify(self, handler):
         form = handler.create_verify_email_form()
         if form.validate_on_submit():
-            return self._send_confirmation(form.email.data)
+            rate_limit_exceeded = not signup_rate_limiter.test()
+            if not rate_limit_exceeded:
+                signup_rate_limiter.hit()
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py:279]:
```suggestion
            if signup_rate_limiter.hit():                
```

File: indico/modules/auth/controllers.py [hunk 2]
```diff
@@ -273,7 +274,15 @@ def _process(self):
...
+                return self._send_confirmation(form.email.data)
+            else:
+                retry_in = signup_rate_limiter.get_reset_delay()
+                delay = format_human_timedelta(retry_in, 'minutes')
+                flash(_('Too many verification email sent. Please wait {}').format(delay), 'warning')
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py:284]:
```suggestion
                flash(_('Too many verification emails sent. Please wait {}').format(delay), 'warning')
```

File: indico/core/config.py [hunk 3]
```diff
@@ -51,6 +51,7 @@
     'EXTERNAL_REGISTRATION_URL': None,
     'HELP_URL': 'https://learn.getindico.io',
     'FAILED_LOGIN_RATE_LIMIT': '5 per 15 minutes; 10 per day',
+    'SIGNUP_RATE_LIMIT': '3 per 2 minutes; 20 per day',
```
Comment (ThiefMaster) [inline indico/core/config.py:54]:
```suggestion
    'SIGNUP_RATE_LIMIT': '3 per 10 minutes; 20 per day',
```

should be more than enough.

also this needs to to be documented in settings.rst (like i did for `FAILED_LOGIN_RATE_LIMIT`). you can copy that text and just adapt the parts that mention logging in etc

**Review by ThiefMaster — CHANGES_REQUESTED** [review]

File: docs/source/config/settings.rst [hunk 4]
```diff
@@ -85,6 +85,23 @@ Authentication
...
+.. data:: SIGNUP_RATE_LIMIT
+
+    Applies a rate limit to sending verification email in signup attempts.
+    When specifying multiple rate limits separated with a semicolon,
+    they are checked in that specific order, which can allow for a short burst of
+    attempts (e.g. a legitimate user trying multiple passwords they commonly use)
+    and then slowing down more strongly (in case someone tries to brute-force more
+    than just a few passwords).
+
+    Rate limiting is applied by IP address and each verification email sent count against the
```
Comment (tomasr8) [inline docs/source/config/settings.rst:97]:
```suggestion
    Rate limiting is applied by IP address and each verification email sent counts against the
```

File: docs/source/config/settings.rst [hunk 5]
```diff
@@ -85,6 +85,23 @@ Authentication
...
     Default: ``'5 per 15 minutes; 10 per day'``
 
+.. data:: SIGNUP_RATE_LIMIT
+
+    Applies a rate limit to sending verification email in signup attempts.
+    When specifying multiple rate limits separated with a semicolon,
+    they are checked in that specific order, which can allow for a short burst of
+    attempts (e.g. a legitimate user trying multiple passwords they commonly use)
+    and then slowing down more strongly (in case someone tries to brute-force more
+    than just a few passwords).
```
Comment (ThiefMaster) [inline docs/source/config/settings.rst:95]: This text doesn't really make sense here...

**Review by ThiefMaster — APPROVED** [review]
i refactored things a bit and also used stricter defaults

**Pull request merged**